// @lunchflow/actual-flow/dist/actual-budget-client.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ActualBudgetClient = void 0;
const actualAPI = __importStar(require("@actual-app/api"));
const fs_1 = __importDefault(require("fs"));
class ActualBudgetClient {
    constructor(serverUrl, budgetSyncId, serverPassword, encryptionPassword) {
        this.connected = false;
        this.serverUrl = serverUrl;
        this.budgetSyncId = budgetSyncId;
        this.serverPassword = serverPassword;
        this.encryptionPassword = encryptionPassword;
        this.dataDir = './actual-data'; // Local cache directory
    }
    async connect() {
        try {
            // Ensure data directory exists
            if (!fs_1.default.existsSync(this.dataDir)) {
                fs_1.default.mkdirSync(this.dataDir, { recursive: true });
            }
            await actualAPI.init({
                dataDir: this.dataDir,
                serverURL: this.serverUrl,
                password: this.serverPassword,
            });
            // Download the budget to local cache with encryption support
            if (this.encryptionPassword && this.encryptionPassword.trim() !== '') {
                await actualAPI.downloadBudget(this.budgetSyncId, {
                    password: this.encryptionPassword,
                });
            }
            else {
                await actualAPI.downloadBudget(this.budgetSyncId);
            }
            this.connected = true;
            return true;
        }
        catch (error) {
            if (error.message.includes('budget directory does not exist')) {
                console.error(`Budget with sync ID "${this.budgetSyncId}" does not exist on the server.`);
                console.error('Please check your budget sync ID and try again.');
            }
            else if (error.message.includes('ECONNREFUSED')) {
                console.error(`Cannot connect to Actual Budget server at ${this.serverUrl}`);
                console.error('Please check that your Actual Budget server is running.');
            }
            else if (error.message.includes('not found')) {
                console.error(`Budget with sync ID "${this.budgetSyncId}" not found on the server.`);
                console.error('Please verify your budget sync ID in Actual Budget settings.');
            }
            else {
                console.error('Failed to connect to Actual Budget:', error.message);
            }
            this.connected = false;
            return false;
        }
    }
    async testConnection() {
        try {
            await this.connect();
            return this.connected;
        }
        catch (error) {
            console.error('Actual Budget connection test failed:', error.message);
            return false;
        }
    }
    async getAccounts() {
        if (!this.connected) {
            await this.connect();
        }
        try {
            const accounts = await actualAPI.getAccounts();
            return accounts.map((account) => ({
                id: account.id,
                name: account.name,
                type: account.type,
                balance: actualAPI.utils.integerToAmount(account.balance || 0),
                currency: 'USD', // Assuming USD, adjust as needed
            }));
        }
        catch (error) {
            console.error('Failed to fetch Actual Budget accounts:', error.message);
            throw new Error(`Failed to fetch accounts: ${error.message}`);
        }
    }
    async getTransactions(accountId) {
        if (!this.connected) {
            await this.connect();
        }
        try {
            let transactions;
            // Use wide date range to get all transactions (API requires non-null dates in v25.12+)
            const startDate = '1970-01-01';
            const endDate = '2099-12-31';
            if (accountId) {
                // Get transactions for a specific account
                transactions = await actualAPI.getTransactions(accountId, startDate, endDate);
            }
            else {
                // Get transactions for all accounts
                const accounts = await this.getAccounts();
                const allTransactions = [];
                for (const account of accounts) {
                    try {
                        const accountTransactions = await actualAPI.getTransactions(account.id, startDate, endDate);
                        allTransactions.push(...accountTransactions);
                    }
                    catch (error) {
                        console.warn(`Failed to fetch transactions for account ${account.name}:`, error.message);
                        // Continue with other accounts
                    }
                }
                transactions = allTransactions;
            }
            return transactions.map((transaction) => ({
                id: transaction.id,
                date: transaction.date,
                amount: transaction.amount,
                imported_payee: transaction.imported_payee || transaction.payee_name || '',
                payee_name: transaction.payee_name || '',
                account: transaction.account,
                cleared: transaction.cleared,
                notes: transaction.notes,
                imported_id: transaction.imported_id,
            }));
        }
        catch (error) {
            console.error('Failed to fetch Actual Budget transactions:', error.message);
            throw new Error(`Failed to fetch transactions: ${error.message}`);
        }
    }
    async importTransactions(transactions) {
        if (!this.connected) {
            await this.connect();
        }
        try {
            // Group transactions by account
            const transactionsByAccount = new Map();
            for (const transaction of transactions) {
                if (!transactionsByAccount.has(transaction.account)) {
                    transactionsByAccount.set(transaction.account, []);
                }
                transactionsByAccount.get(transaction.account).push(transaction);
            }
            // Import transactions for each account
            const importPromises = Array.from(transactionsByAccount.entries()).map(async ([accountId, accountTransactions]) => {
                try {
                    await actualAPI.importTransactions(accountId, accountTransactions);
                    console.log(`Imported ${accountTransactions.length} transactions to account ${accountId}`);
                }
                catch (error) {
                    console.error(`Failed to import transactions to account ${accountId}:`, error.message);
                    throw new Error(`Failed to import transactions to account ${accountId}: ${error.message}`);
                }
            });
            await Promise.all(importPromises);
        }
        catch (error) {
            console.error('Failed to import transactions to Actual Budget:', error.message);
            throw new Error(`Failed to import transactions: ${error.message}`);
        }
    }
    async listAvailableBudgets() {
        try {
            // Ensure data directory exists
            if (!fs_1.default.existsSync(this.dataDir)) {
                fs_1.default.mkdirSync(this.dataDir, { recursive: true });
            }
            await actualAPI.init({
                dataDir: this.dataDir,
                serverURL: this.serverUrl,
                password: this.serverPassword,
            });
            const budgets = await actualAPI.getBudgets();
            return budgets.map((budget) => ({
                id: budget.id,
                name: budget.name || 'Unnamed Budget'
            }));
        }
        catch (error) {
            console.error('Failed to fetch available budgets:', error.message);
            throw new Error(`Failed to fetch budgets: ${error.message}`);
        }
    }
    async shutdown() {
        try {
            await actualAPI.shutdown();
            this.connected = false;
        }
        catch (error) {
            console.error('Error during shutdown:', error.message);
        }
    }
}
exports.ActualBudgetClient = ActualBudgetClient;
