// @lunchflow/actual-flow/dist/config-manager.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConfigManager = void 0;
const fs_1 = __importDefault(require("fs"));
const path_1 = __importDefault(require("path"));
class ConfigManager {
    constructor(configPath = path_1.default.join(process.cwd(), 'config.json')) {
        this.configPath = configPath;
    }
    loadConfig() {
        try {
            if (!fs_1.default.existsSync(this.configPath)) {
                return null;
            }
            const configData = fs_1.default.readFileSync(this.configPath, 'utf8');
            const config = JSON.parse(configData);
            // Validate config structure
            if (!this.validateConfig(config)) {
                console.warn('Invalid config file structure, will recreate');
                return null;
            }
            return config;
        }
        catch (error) {
            console.error('Failed to load config:', error);
            return null;
        }
    }
    saveConfig(config) {
        try {
            // Ensure directory exists
            const dir = path_1.default.dirname(this.configPath);
            if (!fs_1.default.existsSync(dir)) {
                fs_1.default.mkdirSync(dir, { recursive: true });
            }
            fs_1.default.writeFileSync(this.configPath, JSON.stringify(config, null, 2));
        }
        catch (error) {
            console.error('Failed to save config:', error);
            throw error;
        }
    }
    createDefaultConfig() {
        return {
            lunchFlow: {
                apiKey: '',
                baseUrl: 'https://api.lunchflow.com',
            },
            actualBudget: {
                serverUrl: '',
                budgetSyncId: '',
                password: '',
                encryptionPassword: undefined,
                duplicateCheckingAcrossAccounts: false, // Default to off
            },
            accountMappings: [],
        };
    }
    updateLunchFlowConfig(apiKey, baseUrl) {
        const config = this.loadConfig() || this.createDefaultConfig();
        config.lunchFlow = { apiKey, baseUrl };
        this.saveConfig(config);
    }
    updateActualBudgetConfig(serverUrl, budgetSyncId, serverPassword, encryptionPassword, duplicateCheckingAcrossAccounts) {
        const config = this.loadConfig() || this.createDefaultConfig();
        config.actualBudget = {
            serverUrl,
            budgetSyncId,
            password: serverPassword,
            encryptionPassword,
            duplicateCheckingAcrossAccounts: duplicateCheckingAcrossAccounts ?? config.actualBudget.duplicateCheckingAcrossAccounts ?? false
        };
        this.saveConfig(config);
    }
    updateAccountMappings(mappings) {
        const config = this.loadConfig();
        if (config) {
            config.accountMappings = mappings;
            this.saveConfig(config);
        }
    }
    validateConfig(config) {
        return (config &&
            typeof config === 'object' &&
            config.lunchFlow &&
            typeof config.lunchFlow.apiKey === 'string' &&
            typeof config.lunchFlow.baseUrl === 'string' &&
            config.actualBudget &&
            typeof config.actualBudget.serverUrl === 'string' &&
            typeof config.actualBudget.budgetSyncId === 'string' &&
            typeof config.actualBudget.password === 'string' &&
            (typeof config.actualBudget.encryptionPassword === 'string' || config.actualBudget.encryptionPassword === undefined) &&
            (typeof config.actualBudget.duplicateCheckingAcrossAccounts === 'boolean' || config.actualBudget.duplicateCheckingAcrossAccounts === undefined) &&
            Array.isArray(config.accountMappings));
    }
    getConfigPath() {
        return this.configPath;
    }
    // Check if config exists and has required fields
    isConfigured() {
        const config = this.loadConfig();
        return config !== null &&
            config.lunchFlow.apiKey.length > 0 &&
            config.actualBudget.serverUrl.length > 0 &&
            config.actualBudget.budgetSyncId.length > 0 &&
            config.actualBudget.password.length > 0;
    }
    // Get a safe version of config for display (hides sensitive data)
    getSafeConfig() {
        const config = this.loadConfig();
        if (!config)
            return null;
        return {
            lunchFlow: {
                apiKey: config.lunchFlow.apiKey ? '***' + config.lunchFlow.apiKey.slice(-4) : '',
                baseUrl: config.lunchFlow.baseUrl,
            },
            actualBudget: {
                serverUrl: config.actualBudget.serverUrl,
                budgetSyncId: config.actualBudget.budgetSyncId,
                password: config.actualBudget.password ? '***' : '',
                encryptionPassword: config.actualBudget.encryptionPassword ? '***' : undefined,
            },
            accountMappings: config.accountMappings,
        };
    }
}
exports.ConfigManager = ConfigManager;
