// @lunchflow/actual-flow/dist/duplicate-detector.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DuplicateTransactionDetector = void 0;
class DuplicateTransactionDetector {
    constructor(existingTransactions = []) {
        this.existingTransactions = new Map();
        this.buildTransactionMap(existingTransactions);
    }
    buildTransactionMap(transactions) {
        this.existingTransactions.clear();
        for (const transaction of transactions) {
            // Use imported_id as the primary key for duplicate detection
            if (transaction.imported_id) {
                this.existingTransactions.set(transaction.imported_id, transaction);
            }
        }
    }
    checkForDuplicates(newTransactions) {
        return newTransactions.map(transaction => {
            const duplicateInfo = this.checkTransaction(transaction);
            return {
                ...transaction,
                isDuplicate: duplicateInfo.isDuplicate,
                duplicateOf: duplicateInfo.duplicateOf,
            };
        });
    }
    checkTransaction(transaction) {
        if (!transaction.imported_id) {
            return { isDuplicate: false };
        }
        const existingTransaction = this.existingTransactions.get(transaction.imported_id);
        if (existingTransaction) {
            return {
                isDuplicate: true,
                duplicateOf: existingTransaction.id,
                existingTransaction,
            };
        }
        return { isDuplicate: false };
    }
    getDuplicateCount(transactions) {
        return transactions.filter(t => t.isDuplicate).length;
    }
    getUniqueTransactions(transactions) {
        return transactions.filter(t => !t.isDuplicate);
    }
    getDuplicateTransactions(transactions) {
        return transactions.filter(t => t.isDuplicate);
    }
}
exports.DuplicateTransactionDetector = DuplicateTransactionDetector;
