// @lunchflow/actual-flow/dist/lunch-flow-client.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LunchFlowClient = void 0;
const axios_1 = __importDefault(require("axios"));
class LunchFlowClient {
    constructor(apiKey, baseUrl = 'https://api.lunchflow.com') {
        this.maxRetries = 3;
        this.apiKey = apiKey;
        this.client = axios_1.default.create({
            baseURL: baseUrl,
            headers: {
                'x-api-key': apiKey,
                'Content-Type': 'application/json',
            },
            timeout: 60000,
        });
    }
    async retryWithBackoff(operation, operationName, retryCount = 0) {
        try {
            return await operation();
        }
        catch (error) {
            const isRetriableError = this.isRetriableError(error);
            if (retryCount < this.maxRetries && isRetriableError) {
                const backoffDelay = Math.pow(2, retryCount) * 1000; // Exponential backoff: 1s, 2s, 4s
                console.log(`${operationName} failed (attempt ${retryCount + 1}/${this.maxRetries + 1}). Retrying in ${backoffDelay}ms...`);
                console.log(`Error: ${error.message}`);
                await this.sleep(backoffDelay);
                return this.retryWithBackoff(operation, operationName, retryCount + 1);
            }
            // Log final failure
            if (retryCount >= this.maxRetries) {
                console.error(`${operationName} failed after ${this.maxRetries + 1} attempts. Final error: ${error.message}`);
            }
            throw error;
        }
    }
    isRetriableError(error) {
        // Retry on network errors, timeouts, and 5xx server errors
        if (error.code === 'ECONNRESET' || error.code === 'ETIMEDOUT' || error.code === 'ENOTFOUND') {
            return true;
        }
        if (error.response?.status >= 500) {
            return true;
        }
        // Retry on timeout errors
        if (error.code === 'ECONNABORTED' && error.message.includes('timeout')) {
            return true;
        }
        return false;
    }
    sleep(ms) {
        return new Promise(resolve => setTimeout(resolve, ms));
    }
    async testConnection() {
        try {
            return await this.retryWithBackoff(async () => {
                // Try to get accounts as a health check
                const response = await this.client.get('/accounts');
                return response.status === 200;
            }, 'Test Lunch Flow connection');
        }
        catch (error) {
            console.error('Lunch Flow connection test failed after all retries:', error.message);
            return false;
        }
    }
    async getAccounts() {
        return this.retryWithBackoff(async () => {
            const response = await this.client.get('/accounts');
            // Handle different possible response structures
            if (Array.isArray(response.data.accounts)) {
                return response.data.accounts;
            }
            else {
                console.warn('Unexpected response structure from Lunch Flow accounts endpoint');
                return [];
            }
        }, 'Fetch Lunch Flow accounts');
    }
    async getTransactions(accountId, includePending = false) {
        return this.retryWithBackoff(async () => {
            const params = includePending ? '?include_pending=true' : '';
            const response = await this.client.get(`/accounts/${accountId}/transactions${params}`);
            // Handle different possible response structures
            if (Array.isArray(response.data.transactions)) {
                return response.data.transactions;
            }
            else {
                console.warn('Unexpected response structure from Lunch Flow transactions endpoint');
                return [];
            }
        }, `Fetch transactions for account ${accountId}`);
    }
}
exports.LunchFlowClient = LunchFlowClient;
