// @lunchflow/actual-flow/dist/ui.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TerminalUI = void 0;
const inquirer_1 = __importDefault(require("inquirer"));
const chalk_1 = __importDefault(require("chalk"));
const ora_1 = __importDefault(require("ora"));
const cli_table3_1 = __importDefault(require("cli-table3"));
class TerminalUI {
    async showWelcome() {
        console.clear();
        console.log(chalk_1.default.blue.bold('\n🍽️  Lunch Flow → Actual Budget Importer\n'));
        console.log(chalk_1.default.gray('This tool helps you import transactions from Lunch Flow to Actual Budget\n'));
    }
    async getLunchFlowCredentials() {
        console.log(chalk_1.default.yellow('📡 Lunch Flow Configuration\n'));
        console.log(chalk_1.default.gray('To find your Lunch Flow API key:'));
        console.log(chalk_1.default.gray('1. Login to Lunch Flow: https://lunchflow.app'));
        console.log(chalk_1.default.gray('2. Go to Destinations → create a new API destination'));
        console.log(chalk_1.default.gray('3. Look for "API Key" - that\'s your API key\n'));
        const answers = await inquirer_1.default.prompt([
            {
                type: 'password',
                name: 'apiKey',
                message: 'Enter your Lunch Flow API key:',
                validate: (input) => input.length > 0 || 'API key is required',
                mask: '*',
            },
            {
                type: 'input',
                name: 'baseUrl',
                message: 'Enter Lunch Flow API base URL:',
                default: 'https://lunchflow.app/api/v1',
                validate: (input) => {
                    try {
                        new URL(input);
                        return true;
                    }
                    catch {
                        return 'Please enter a valid URL';
                    }
                },
            },
        ]);
        return answers;
    }
    async getActualBudgetCredentials() {
        console.log(chalk_1.default.yellow('\n💰 Actual Budget Configuration\n'));
        console.log(chalk_1.default.gray('To find your budget sync ID:'));
        console.log(chalk_1.default.gray('1. Open Actual Budget in your browser'));
        console.log(chalk_1.default.gray('2. Go to Settings → Show advanced settings'));
        console.log(chalk_1.default.gray('3. Look for "Sync ID" - that\'s your budget sync ID\n'));
        console.log(chalk_1.default.gray('You\'ll need two different passwords:'));
        console.log(chalk_1.default.gray('• Server password: Used to connect to your Actual Budget server'));
        console.log(chalk_1.default.gray('• Encryption password: Used for end-to-end encryption (optional)\n'));
        const answers = await inquirer_1.default.prompt([
            {
                type: 'input',
                name: 'serverUrl',
                message: 'Enter Actual Budget server URL:',
                default: 'http://localhost:5006',
                validate: (input) => {
                    try {
                        new URL(input);
                        return true;
                    }
                    catch {
                        return 'Please enter a valid URL';
                    }
                },
            },
            {
                type: 'input',
                name: 'budgetSyncId',
                message: 'Enter Actual Budget budget sync ID:',
                validate: (input) => input.length > 0 || 'Budget sync ID is required',
            },
            {
                type: 'password',
                name: 'password',
                message: 'Enter Actual Budget server password:',
                mask: '*',
                validate: (input) => input.length > 0 || 'Server password is required',
            },
            {
                type: 'password',
                name: 'encryptionPassword',
                message: 'Enter encryption password (leave empty if no end-to-end encryption):',
                mask: '*',
                validate: (input) => true, // Encryption password is optional
            },
            {
                type: 'confirm',
                name: 'duplicateCheckingAcrossAccounts',
                message: 'Enable duplicate checking across all accounts? (prevents importing transactions that already exist in other accounts)',
                default: false,
            },
        ]);
        return answers;
    }
    async showConnectionStatus(status) {
        console.log(chalk_1.default.blue('\n🔗 Connection Status\n'));
        const lfStatus = status.lunchFlow ? chalk_1.default.green('✅ Connected') : chalk_1.default.red('❌ Disconnected');
        const abStatus = status.actualBudget ? chalk_1.default.green('✅ Connected') : chalk_1.default.red('❌ Disconnected');
        console.log(`Lunch Flow: ${lfStatus}`);
        console.log(`Actual Budget: ${abStatus}\n`);
    }
    async showAccountsTable(accounts, title) {
        console.log(chalk_1.default.blue(`\n${title}\n`));
        if (accounts.length === 0) {
            console.log(chalk_1.default.yellow('No accounts found.\n'));
            return;
        }
        const table = new cli_table3_1.default({
            head: ['ID', 'Name'],
            colWidths: [8, 25],
            style: {
                head: ['cyan'],
                border: ['gray'],
            }
        });
        accounts.forEach(account => {
            table.push([
                account.id.toString().substring(0, 8) + '...',
                account.name,
            ]);
        });
        console.log(table.toString());
    }
    async configureAccountMappings(lfAccounts, abAccounts, existingMappings = []) {
        console.log(chalk_1.default.yellow('\n📋 Configure Account Mappings\n'));
        console.log(chalk_1.default.gray('Map each Lunch Flow account to an Actual Budget account:\n'));
        const mappings = [];
        for (const lfAccount of lfAccounts) {
            const existingMapping = existingMappings.find(m => m.lunchFlowAccountId === lfAccount.id);
            if (existingMapping) {
                const currentInfo = `Currently mapped to: ${existingMapping.actualBudgetAccountName}` +
                    (existingMapping.syncStartDate ? ` | Sync start: ${existingMapping.syncStartDate}` : '') +
                    ` | Pending: ${existingMapping.includePending ? 'Yes' : 'No'}`;
                const { action } = await inquirer_1.default.prompt([{
                        type: 'list',
                        name: 'action',
                        message: `"${lfAccount.name}" — ${currentInfo}\n  What would you like to do?`,
                        choices: [
                            { name: 'Keep as is', value: 'keep' },
                            { name: 'Edit mapping', value: 'edit' },
                            { name: 'Delete mapping', value: 'delete' },
                        ],
                    }]);
                if (action === 'keep') {
                    mappings.push(existingMapping);
                    continue;
                }
                if (action === 'delete') {
                    continue;
                }
                // action === 'edit': fall through to account-selection prompts
            }
            const choices = abAccounts.map(abAccount => ({
                name: `${abAccount.name} (${abAccount.currency})`,
                value: abAccount.id,
            }));
            const answer = await inquirer_1.default.prompt([
                {
                    type: 'list',
                    name: 'abAccountId',
                    message: `Map "${lfAccount.name}" (${lfAccount.institution_name}) to:`,
                    default: existingMapping?.actualBudgetAccountId ?? 'skip',
                    choices: [
                        { name: 'Skip this account', value: 'skip' },
                        ...choices,
                    ],
                },
            ]);
            if (answer.abAccountId !== 'skip') {
                const abAccount = abAccounts.find(a => a.id === answer.abAccountId);
                if (abAccount) {
                    // Ask for optional sync start date
                    const dateAnswer = await inquirer_1.default.prompt([
                        {
                            type: 'input',
                            name: 'syncStartDate',
                            message: `Sync start date for "${lfAccount.name}" (YYYY-MM-DD, optional - press Enter to skip):`,
                            default: existingMapping?.syncStartDate ?? '',
                            validate: (input) => {
                                if (!input.trim())
                                    return true; // Empty is allowed
                                const dateRegex = /^\d{4}-\d{2}-\d{2}$/;
                                if (!dateRegex.test(input)) {
                                    return 'Please enter date in YYYY-MM-DD format or leave empty';
                                }
                                const date = new Date(input);
                                if (isNaN(date.getTime())) {
                                    return 'Please enter a valid date';
                                }
                                return true;
                            },
                        },
                    ]);
                    // Ask whether to include pending transactions
                    const pendingAnswer = await inquirer_1.default.prompt([
                        {
                            type: 'confirm',
                            name: 'includePending',
                            message: `Include pending (unposted) transactions for "${lfAccount.name}"?`,
                            default: existingMapping?.includePending ?? false,
                        },
                    ]);
                    const mapping = {
                        lunchFlowAccountId: lfAccount.id,
                        lunchFlowAccountName: lfAccount.name,
                        actualBudgetAccountId: abAccount.id,
                        actualBudgetAccountName: abAccount.name,
                    };
                    if (dateAnswer.syncStartDate.trim()) {
                        mapping.syncStartDate = dateAnswer.syncStartDate.trim();
                    }
                    if (pendingAnswer.includePending) {
                        mapping.includePending = true;
                    }
                    mappings.push(mapping);
                }
            }
        }
        return mappings;
    }
    async showAccountMappings(mappings) {
        console.log(chalk_1.default.blue('\n📋 Current Account Mappings\n'));
        if (mappings.length === 0) {
            console.log(chalk_1.default.yellow('No account mappings configured.\n'));
            return;
        }
        const table = new cli_table3_1.default({
            head: ['Lunch Flow Account', '→', 'Actual Budget Account', 'Sync Start', 'Pending'],
            colWidths: [25, 3, 25, 12, 10],
            style: {
                head: ['cyan'],
                border: ['gray'],
            }
        });
        mappings.forEach(mapping => {
            table.push([
                mapping.lunchFlowAccountName,
                '→',
                mapping.actualBudgetAccountName,
                mapping.syncStartDate || chalk_1.default.gray('None'),
                mapping.includePending ? chalk_1.default.green('Yes') : chalk_1.default.gray('No')
            ]);
        });
        console.log(table.toString());
    }
    async confirmImport(transactionCount, dateRange) {
        console.log(chalk_1.default.yellow('\n⚠️  Import Confirmation\n'));
        console.log(`Date Range: ${dateRange.startDate} to ${dateRange.endDate}`);
        console.log(`Transactions to import: ${transactionCount}\n`);
        const answer = await inquirer_1.default.prompt([
            {
                type: 'confirm',
                name: 'confirm',
                message: 'Proceed with import?',
                default: true,
            },
        ]);
        return answer.confirm;
    }
    async showMainMenu() {
        console.log(chalk_1.default.blue('\n🎯 Main Menu\n'));
        console.log(chalk_1.default.gray('💡 Tip: Use "actual-flow import" to run import directly\n'));
        const answer = await inquirer_1.default.prompt([
            {
                type: 'list',
                name: 'action',
                message: 'What would you like to do?',
                choices: [
                    { name: '🔗 Test connections', value: 'test' },
                    { name: '📋 List available budgets', value: 'list-budgets' },
                    { name: '📋 Configure account mappings', value: 'configure' },
                    { name: '📊 Show current mappings', value: 'show' },
                    { name: '📥 Import transactions', value: 'import' },
                    { name: '⚙️  Reconfigure credentials', value: 'reconfigure' },
                    { name: '❌ Exit', value: 'exit' },
                ],
            },
        ]);
        return answer.action;
    }
    async showReconfigureMenu() {
        console.log(chalk_1.default.yellow('\n⚙️  Reconfigure Credentials\n'));
        const answer = await inquirer_1.default.prompt([
            {
                type: 'list',
                name: 'action',
                message: 'What would you like to reconfigure?',
                choices: [
                    { name: 'Lunch Flow credentials', value: 'lunchflow' },
                    { name: 'Actual Budget credentials', value: 'actualbudget' },
                    { name: 'Both', value: 'both' },
                    { name: 'Cancel', value: 'cancel' },
                ],
            },
        ]);
        return answer.action;
    }
    showSpinner(message) {
        return (0, ora_1.default)(message).start();
    }
    showSuccess(message) {
        console.log(chalk_1.default.green(`✅ ${message}`));
    }
    showError(message) {
        console.log(chalk_1.default.red(`❌ ${message}`));
    }
    showInfo(message) {
        console.log(chalk_1.default.blue(`ℹ️  ${message}`));
    }
    showWarning(message) {
        console.log(chalk_1.default.yellow(`⚠️  ${message}`));
    }
    async showTransactionPreview(transactions, accounts, count = 10) {
        console.log(chalk_1.default.blue(`\n📊 Transaction Preview (showing first ${Math.min(count, transactions.length)})\n`));
        if (transactions.length === 0) {
            console.log(chalk_1.default.yellow('No transactions to preview.\n'));
            return;
        }
        const accountNames = accounts.reduce((acc, account) => {
            acc[account.id] = account.name;
            return acc;
        }, {});
        // Count duplicates for summary
        const duplicateCount = transactions.filter(t => t.isDuplicate).length;
        const uniqueCount = transactions.length - duplicateCount;
        if (duplicateCount > 0) {
            console.log(chalk_1.default.yellow(`⚠️  Found ${duplicateCount} duplicate transactions out of ${transactions.length} total\n`));
        }
        const table = new cli_table3_1.default({
            head: ['Date', 'Description', 'Amount', 'Account', 'Status'],
            colWidths: [12, 25, 12, 15, 10],
            style: {
                head: ['cyan'],
                border: ['gray'],
            }
        });
        transactions
            .sort((a, b) => new Date(b.date).getTime() - new Date(a.date).getTime())
            .slice(0, count)
            .forEach(transaction => {
            const amount = transaction.amount / 100;
            const amountDisplay = transaction.amount >= 0
                ? chalk_1.default.green(`+${amount.toFixed(2)}`)
                : chalk_1.default.red(`-${Math.abs(amount).toFixed(2)}`);
            const status = transaction.isDuplicate
                ? chalk_1.default.red('DUPLICATE')
                : chalk_1.default.green('NEW');
            const description = transaction.isDuplicate
                ? chalk_1.default.gray(transaction.imported_payee)
                : transaction.imported_payee;
            table.push([
                transaction.date,
                description,
                amountDisplay,
                accountNames[transaction.account] || 'Unknown',
                status
            ]);
        });
        console.log(table.toString());
        if (transactions.length > count) {
            console.log(chalk_1.default.gray(`... and ${transactions.length - count} more transactions\n`));
        }
        else {
            console.log();
        }
        // Show summary
        if (duplicateCount > 0) {
            console.log(chalk_1.default.yellow(`📋 Summary: ${uniqueCount} new transactions, ${duplicateCount} duplicates (will be skipped)\n`));
        }
    }
}
exports.TerminalUI = TerminalUI;
